### ISSUE TITLE
Lisp error: (json-readtable-error): Flycheck should error out better if needed binary isn't present.

### ISSUE THREAD

**Issue by stieg** [flycheck/flycheck#1312]
I hit the aforementioned error while using Flycheck with Typscript.  Problem was that typescript binary was not available on the PATH.  Thus it would crash like this:
```
Debugger entered--Lisp error: (json-readtable-error)
  signal(json-readtable-error nil)
  json-read()
  json-read-from-string("module.js:491\n    throw err;\n    ^\n\nError: Cannot find module 'typescript'\n    at Function.Module._resolveFilename (module.js:489:15)\n    at Function.Module._load (module.js:439:25)\n    at Module.require (module.js:517:17)\n    at require (internal/module.js:11:18)\n    at Object.<anonymous> (/usr/lib/node_modules/tslint/lib/linter.js:20:10)\n    at Module._compile (module.js:573:30)\n    at Object.Module._extensions..js (module.js:584:10)\n    at Module.load (module.js:507:32)\n    at tryModuleLoad (module.js:470:12)\n    at Function.Module._load (module.js:462:3)\n")
  flycheck-parse-tslint("module.js:491\n    throw err;\n    ^\n\nError: Cannot find module 'typescript'\n    at Function.Module._resolveFilename (module.js:489:15)\n    at Function.Module._load (module.js:439:25)\n    at Module.require (module.js:517:17)\n    at require (internal/module.js:11:18)\n    at Object.<anonymous> (/usr/lib/node_modules/tslint/lib/linter.js:20:10)\n    at Module._compile (module.js:573:30)\n    at Object.Module._extensions..js (module.js:584:10)\n    at Module.load (module.js:507:32)\n    at tryModuleLoad (module.js:470:12)\n    at Function.Module._load (module.js:462:3)\n" typescript-tslint #<buffer user-service.ts>)
  flycheck-parse-output("module.js:491\n    throw err;\n    ^\n\nError: Cannot find module 'typescript'\n    at Function.Module._resolveFilename (module.js:489:15)\n    at Function.Module._load (module.js:439:25)\n    at Module.require (module.js:517:17)\n    at require (internal/module.js:11:18)\n    at Object.<anonymous> (/usr/lib/node_modules/tslint/lib/linter.js:20:10)\n    at Module._compile (module.js:573:30)\n    at Object.Module._extensions..js (module.js:584:10)\n    at Module.load (module.js:507:32)\n    at tryModuleLoad (module.js:470:12)\n    at Function.Module._load (module.js:462:3)\n" typescript-tslint #<buffer user-service.ts>)
  flycheck-finish-checker-process(typescript-tslint 1 ("/tmp/flycheck1344811g/user-service.ts" "/tmp/flycheck1344811g") "module.js:491\n    throw err;\n    ^\n\nError: Cannot find module 'typescript'\n    at Function.Module._resolveFilename (module.js:489:15)\n    at Function.Module._load (module.js:439:25)\n    at Module.require (module.js:517:17)\n    at require (internal/module.js:11:18)\n    at Object.<anonymous> (/usr/lib/node_modules/tslint/lib/linter.js:20:10)\n    at Module._compile (module.js:573:30)\n    at Object.Module._extensions..js (module.js:584:10)\n    at Module.load (module.js:507:32)\n    at tryModuleLoad (module.js:470:12)\n    at Function.Module._load (module.js:462:3)\n" #[128 "\301\302\300#\207" [[cl-struct-flycheck-syntax-check #<buffer user-service.ts> typescript-tslint #<process flycheck-typescript-tslint> "<REDACTED>"] apply flycheck-report-buffer-checker-status] 5 "\n\n(fn &rest ARGS)"] "<REDACTED>")
  flycheck-handle-signal(#<process flycheck-typescript-tslint> "exited abnormally with code 1\n")
```
Installing typescript globally solved the issue.  But from reading the error message displayed on the bottom bar it was very not clear that is what the issue was.  Nor did the requirements state that typescript needed to be in the path (it could be installed within the node_modules directory).  So there are two issues here:

* Improving the error message when binary is not found on the PATH
* Improving documentation to note that typescript must be on the PATH (unless that is not intended).

**Comment by cpitclaudel** [comment]
Thanks for the report! Sorry for the trouble. Can you also post the missing bits, primarily the output C-c ! v and a confirmation that this error also happens in emacs -Q?

It'll be easier to diagnose the issue with that extra info, but I'm not too sure that it's a PATH issue.  The backtrace above suggests that Flycheck did manage to run *something*, and that something produced unexpected output.

Could you also try to run C-c ! C-c, select typescript-tslint, and see what program is being run?

Thanks!

**Comment by stieg** [comment]
> Thanks for the report! Sorry for the trouble.

No problem and no trouble when compared to what you are providing for the emacs community.  Thank you very much for maintaining it.

Here is the output from flycheck-verify-setup:
```
Syntax checkers for buffer auto-complete-request.ts in typescript-mode:

  typescript-tide
    - may enable:        yes
    - predicate:         t
    - Typescript server: running
    - Tide mode:         enabled

  typescript-tslint
    - may enable:         yes
    - executable:         Found at /usr/bin/tslint
    - configuration file: Found at "/home/stieg/Devel/<REDACTED>/tslint.json"

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 30
Emacs version:    25.2.1
System:           x86_64-redhat-linux-gnu
Window system:    nil
```
When I run `emacs -Q`, open up a .ts file and set major mode typescript and enable flycheck I still see the same issue:
```
Syntax checkers for buffer auto-complete-request.ts in typescript-mode:

  typescript-tslint
    - may enable:         yes
    - executable:         Found at /usr/bin/tslint
    - configuration file: Found at "/home/stieg/Devel/<REDACTED>/tslint.json"

Flycheck Mode is enabled.  Use C-u C-c ! x to enable disabled
checkers.

--------------------

Flycheck version: 30
Emacs version:    25.2.1
System:           x86_64-redhat-linux-gnu
Window system:    nil
```

Ran C-c ! C-c with typescript-tslint.  Here is what I saw:

```
-*- mode: compilation; default-directory: "~/Devel/<REDACTED>/" -*-
Compilation started at Wed Aug 30 08:00:48

tslint --format json --config /home/stieg/Devel/<REDACTED>/tslint.json /home/stieg/Devel/<REDACTED>/auto-complete-request.ts
module.js:491
    throw err;
    ^

Error: Cannot find module 'typescript'
    at Function.Module._resolveFilename (module.js:489:15)
    at Function.Module._load (module.js:439:25)
    at Module.require (module.js:517:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/usr/lib/node_modules/tslint/lib/linter.js:20:10)
    at Module._compile (module.js:573:30)
    at Object.Module._extensions..js (module.js:584:10)
    at Module.load (module.js:507:32)
    at tryModuleLoad (module.js:470:12)
    at Function.Module._load (module.js:462:3)

Compilation exited abnormally with code 1 at Wed Aug 30 08:00:48
```

Yeah... that is funny.  tslint is trying to use typescript but its not there.  Lets look at the package.json to see why:

```json
{
  "dependencies": {
    "babel-code-frame": "^6.22.0",
    "colors": "^1.1.2",   
    "commander": "^2.9.0",
    "diff": "^3.2.0",
    "glob": "^7.1.1",
    "minimatch": "^3.0.4",
    "resolve": "^1.3.2",
    "semver": "^5.3.0",
    "tslib": "^1.7.1",
    "tsutils": "^2.5.1"
  },
  "deprecated": false,
  "description": "An extensible static analysis linter for the TypeScript language",
  "devDependencies": {
    "@types/babel-code-frame": "^6.20.0",
    "@types/chai": "^3.5.0",
    "@types/colors": "^1.1.3",   
    "@types/commander": "^2.9.0",
    "@types/diff": "^3.2.0",
    "@types/glob": "^5.0.30",
    "@types/js-yaml": "^3.5.31",
    "@types/minimatch": "^2.0.29",
    "@types/mocha": "^2.2.35",
    "@types/node": "^7.0.29",
    "@types/resolve": "^0.0.4",
    "@types/semver": "^5.3.30",
    "chai": "^3.5.0",
    "github": "^8.2.1",
    "js-yaml": "^3.7.0",
    "json-stringify-pretty-compact": "^1.0.3",
    "mocha": "^3.2.0",
    "npm-run-all": "^4.0.2",
    "nyc": "^10.2.0",
    "rimraf": "^2.5.4",
    "tslint": "^5.4.2",
    "tslint-test-config-non-relative": "file:test/external/tslint-test-config-non-relative",
    "typescript": "2.1"
  }
}
```
Huh.... typescript is a dev dependency here.  I am curious if that is the issue.  Hope that helps.

**Comment by fmdkdd** [comment]
cc @Simplify

**Comment by Simplify** [comment]
@stieg Sorry for late reply.

It seems that you have both `tslint` and `typescript` as devDependencies but you are using `tslint` from `/usr/bin/tslint` (globally installed?) and that may be different version than that in devDependencies. 

Globally installed package will always search for global installations of other packages that they need. For example `eslint` with configuration that require `eslint-plugin-react` will complain if that package is also not installed globally, even if package is installed in your project's node_modules. `eslint` just can't see it (limitation of node/npm/whatever?).

You should set you PATH to point to `./node_modules/.bin/` before normal `$PATH`. You can do it with `.dir-locals.el`, for example. That will make sure that you really use stuff declared in devDependencies and that versions in your package.json are correct for other developers on your project.

I'm not sure what we can do to improve experience in this case.

**Comment by stieg** [comment]
@Simplify I'm not quite sure I groked all of your reply.  Let me break it down and see if we can find ourselves on the same page:

> Globally installed package will always search for global installations of other packages that they need.

Based on the way the Javascript package loader works this is indeed true (recursively scanning the parent node_modules folder for the package it needs until it hits root of the file system; or something close to this anyways).  I'm not arguing this point.

When I referenced the `package.json` file above, it was the one provided for the globally installed `tslint` package.  What I found entertaining was that the `tslint` package listed `typescript` as a devDependency and not a straight up standard dependency.  As we can see `tslint` can not function properly without it (as is noted in the stack I provided above).  Thus in my mind it (typescript) absolutely should be listed under the standard dependency section for just this reason.  What I couldn't figure out is why `typescript` didn't get installed in the first place (I don't recall specifying the --production flag to NPM which would omit installation of devDependencies)

> I'm not sure what we can do to improve experience in this case.

Ahhh... now this is the fun part.  The main issue this ticket tracks is that the error message I received when I hit this issue originally was extremely obtuse and unhelpful.  It read something along the lines of:
```
json-readtable-error
```
when the real issue was that `typescript` module could not be found.   I suppose I should have made this more clear in my initial report.  Sorry about that.

As we can see from the stack I included at the top the error message makes it all the way up to module.js:491.  Then it seems to get lost in the json-read method (perhaps swallowed somehow)?  My suggestion would be to have the error message bubble up to the top so that instead of the generic `json-readtable-error` that I received, I would instead receive a more helpful message like the one that the error described.  That is how I would attempt to remedy this situation.

**Comment by cpitclaudel** [comment]
This was fixed in #1211, which was released after (but, ironically, committed after) this issue was open.  The tslint checker now only parses lines that look like json, so you'd get a Flycheck message saying that no issues were found in the output despite a non-zero error code, and that would include the full error from the checker.
